service/controller/cluster/handler/clusterreleaseversion/checkupgradeprogress.go

```diff
@@ -56,72 +92,52 @@ func (r *Resource) isUpgradeCompleted(ctx context.Context, cluster *capi.Cluster
-			// A node pool is being created, this is the case for first upgrade
-			// to node pools release, as cluster upgrade will trigger first node
-			// pool creation.
-			allNodePoolsUpgraded = false
-			break
+		if !isUpgradedToDesiredReleaseVersion(&machinePool, desiredClusterReleaseVersion) {
+			r.logger.Debugf(ctx, "MachinePool %s has not been upgraded yet", machinePool.Name)
+			return false, nil
 		}
+	}
 
-		if conditions.IsUpgradingTrue(&machinePool) {
-			// A node pool is being upgraded.
-			allNodePoolsUpgraded = false
-			break
-		}
+	r.logger.Debugf(ctx, "node pools upgrade has been completed")
+	return true, nil
+}
 
-		desiredMachinePoolReleaseVersion := key.ReleaseVersion(&machinePool)
-		if desiredMachinePoolReleaseVersion != desiredClusterReleaseVersion {
-			// A node pool upgrade has not been started yet.
-			allNodePoolsUpgraded = false
-			break
-		}
+func isUpgradedToDesiredReleaseVersion(obj conditions.Object, desiredClusterReleaseVersion string) bool {
+	if conditions.IsCreatingTrue(obj) {
+		// Keeping this for backward compatibility, but it should not be necessary:
+		// We are in the upgrade process, but a node pool is being created. This
+		// is the case for first upgrade to node pools release, so an upgrade
+		// from v12.x.x to v13.x.x, as cluster upgrade will trigger first node
+		// pool creation.
+		return false
+	}
 
-		machinePoolLastDeployedReleaseVersion, machinePoolLastDeployedReleaseVersionSet := machinePool.Annotations[annotation.LastDeployedReleaseVersion]
-		if !machinePoolLastDeployedReleaseVersionSet {
-			// A node pool is still not created. This should be caught above in
-			// Creating check, but let's err on the side of caution here.
-			allNodePoolsUpgraded = false
-			break
-		}
+	if conditions.IsUpgradingTrue(obj) {
+		// Control plane node or node pool is being upgraded.
+		return false
+	}
 
-		if machinePoolLastDeployedReleaseVersion != desiredClusterReleaseVersion {
-			// A node pool has not yet been upgraded to the desired release version.
-			allNodePoolsUpgraded = false
-			break
-		}
+	currentDesiredReleaseVersion := key.ReleaseVersion(obj)
+	if currentDesiredReleaseVersion != desiredClusterReleaseVersion {
+		// A control plane node or node pool upgrade has not been started yet.
+		return false
 	}
 
-	// (1) Cluster became ready during the upgrade and the upgrade has been
-	// completed for all cluster descendants.
-	// This is basically upgrade happy path.
-	clusterIsReadyAndUpgraded := becameReadyWhileUpgrading && allNodePoolsUpgraded
-	if clusterIsReadyAndUpgraded {
-		r.logger.Debugf(ctx, "cluster became ready during the upgrade, all cluster descendants are upgraded, upgrade completed")
-		return true, nil
+	lastDeployedReleaseVersion, ok := obj.GetAnnotations()[annotation.LastDeployedReleaseVersion]
+	if !ok {
+		// Control plane or a node pool is still not created. This should be
+		// caught above in Creating check, but let's err on the side of caution
+		// here.
```

Comment:
is this because the annotation is added when `MachinePool` is `Ready`?

Reply:
The annotation would be added for the first time when the control plane or node pool is created, and in the case of MachinePool, I believe that it checks the nodes directly (not Ready condition).

If the annotation is not set, that case would be probably caught in the `IsCreatingTrue(obj)` check above, because if the creation is still in progress (Creating==True), the annotation is not set.

That being said, it could happen that we reorder the handlers, or due to some async nature, so that Creating condition would still not be set, so `IsCreatingTrue(...)` would return false (rare case, but not impossible). Hence this additional check if the annotation is actually set.

P.S. If the annotation is not set, the bellow `lastDeployedReleaseVersion != desiredClusterReleaseVersion` check would also return false, but I think I added this before so it's more explicit.

Reply:
The story around cluster upgrades, labels and annotation used for versioning and upgrades, and also conditions around those things is due for a nice refactoring and making a nicer and cleaner story there, since now some parts are a bit weirdly entangled :)

Reply:
This PR is also towards that untangling :)